source/common/http/alternate_protocols.h

```diff
@@ -0,0 +1,103 @@
+class AlternateProtocols {
+public:
+  // Represents an HTTP origin to be connected too.
+  struct Origin {
+  public:
+    Origin(absl::string_view scheme, absl::string_view hostname, int port);
+
+    bool operator==(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) ==
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator!=(const Origin& other) const { return !this->operator==(other); }
+
+    bool operator<(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) <
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator>(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) >
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator<=(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) <=
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator>=(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) >=
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    std::string scheme_;
+    std::string hostname_;
+    int port_{};
+  };
+
+  // Represents an alternative protocol that can be used to connect to an origin.
+  struct AlternateProtocol {
+  public:
+    AlternateProtocol(absl::string_view alpn, absl::string_view hostname, int port);
+
+    bool operator==(const AlternateProtocol& other) const {
+      return std::tie(alpn_, hostname_, port_) ==
+             std::tie(other.alpn_, other.hostname_, other.port_);
+    }
+
+    bool operator!=(const AlternateProtocol& other) const { return !this->operator==(other); }
+
+    std::string alpn_;
+    std::string hostname_;
+    int port_;
+  };
+
+  explicit AlternateProtocols(TimeSource& time_source);
+
+  // Sets the possible alternative protocols which can be used to connect to the
+  // specified origin. Expires after the specified expiration time.
+  void setAlternatives(const Origin& origin, const std::vector<AlternateProtocol>& protocols,
+                       const MonotonicTime& expiration);
+
+  // Returns the possible alternative protocols which can be used to connect to the
+  // specified origin, or nullptr if not alternatives are found. The returned pointer
+  // is owned by the AlternateProtocols and is valid until the next operation on
+  // AlternateProtocols.
+  const std::vector<AlternateProtocol>* findAlternatives(const Origin& origin);
+
+  // Returns the number of entries in the map.
+  size_t size() const;
+
+private:
+  struct Entry {
+    std::vector<AlternateProtocol> protocols_;
+    MonotonicTime expiration_;
+  };
+
+  // Time source used to check expiration of entries.
+  TimeSource& time_source_;
+
+  // Map from hostname to list of alternate protocols.
+  // TODO(RyanTheOptimist): Add a limit to the size of this map and evict based on usage.
```

Comment:
"Hey have you done any bounds limits on -- Excellent, there's the TODO I was looking for"